Add tests for S2I Perl and Python images with Hot Deploy

mnagy opened the pull request:
@jhadvig

mnagy commented:
I've decided to add in Python as well since it depends on the helper functions.

[test][extended:core]

jhadvig commented on test/extended/images/helper.go:
```diff
@@ -16,3 +24,52 @@ func GetImageLabels(c client.ImageStreamImageInterface, imageRepoName, imageRef
+}
+
+// GetPageCount makes a http request for an example application and returns the page count
+func GetPageCount(oc *exutil.CLI, endpoint string, re *regexp.Regexp) (int, error) {
```
Would it be better to swap the regexp with a string ? Dont se any benefit of having a regular here, since we would like to reuse this method also for other framework

mnagy replied:
Yes, will do.

jhadvig commented on test/extended/images/helper.go:
```diff
@@ -16,3 +24,52 @@ func GetImageLabels(c client.ImageStreamImageInterface, imageRepoName, imageRef
 }
+
+// ModifySourceCode will modify source code in the pod of the application
+// according to the sed script.
```
s/sed script/sed expression/

mnagy replied:
I think it doesn't matter, either one is fine. From sed(1):
```
       -e script, --expression=script

              add the script to the commands to be executed
```

mfojtik commented on test/extended/images/helper.go:
```diff
@@ -16,3 +24,52 @@ func GetImageLabels(c client.ImageStreamImageInterface, imageRepoName, imageRef
+	if err != nil {
+		return err
+	}
+	return oc.Run("exec").Args(pod.Name, "-c", pod.Spec.Containers[0].Name, "--", "sed", "-ie", sedScript, file).Execute()
```
don't need to use container name, if you specify a pod it should pick the first container by default AFAIK

mnagy replied:
I know, but you will get a warning, which is annoying.

jhadvig replied:
Now when I think about it, wouldnt it be better to be able to run any kind of command here ? and not just sticking to sed ?

bparees replied:
@jhadvig let's cross that bridge when we have a use case.

jhadvig replied:
@bparees I've already have one(deleting `public/index.html` in rails), but sure, I can update it in the following PR I have prepared.

mfojtik replied:
if you have that use-case, wait for this to merge and then change the function to work the way you want.

mnagy commented:
Addressed comments, fixed Python tests and rebased for good measure.

PTAL

mnagy commented:
Python and Perl hot deploy test work, search Jenkins logs for "hot deploy".

jhadvig commented on test/extended/images/helper.go:
```diff
@@ -1,12 +1,20 @@
+	"k8s.io/kubernetes/pkg/labels"
 )
 
-// GetImageLabels retrieves Docker labels from image from image repository name and
```
Is this change necessary ?

mnagy replied:
No. No idea how that change got there, fixed.

mfojtik commented on test/extended/images/helper.go:
```diff
@@ -16,3 +24,48 @@ func GetImageLabels(c client.ImageStreamImageInterface, imageRepoName, imageRef
+	}
+
+	// Make the request
+	resp, err := http.Get(fmt.Sprintf("http://%s/", address))
```
I think @bparees has some function to query the HTTP endpoint already for example tests.

mfojtik replied:
also I think this need some retry logic

mnagy replied:
Why retry logic? The pod should be ready at this point. Problem with retrying would be that the count could be different than expected.

bparees replied:
https://github.com/openshift/origin/blob/master/test/extended/util/framework.go#L446

mfojtik replied:
Does the Pod have readiness check that makes sure the pod is "ready" ?

bparees replied:
https://github.com/openshift/origin/blob/master/test/extended/images/sample_repos.go#L63-L74

mnagy replied:
@mfojtik yes. CheckPageContains() calls GetEndpointAddress() which calls WaitForAnEndpoint(). That is equivalent to the code @bparees linked above.

bparees replied:
@mnagy you should definitely be wrappering the existing function here.  if you want to add response code checking to the existing function, that's fine.

mnagy replied:
Ack, will do.

mnagy commented:
@mfojtik addressed your comments and rebased.

bparees commented on test/extended/images/helper.go:
```diff
@@ -16,3 +24,48 @@ func GetImageLabels(c client.ImageStreamImageInterface, imageRepoName, imageRef
+
+// ModifySourceCode will modify source code in the pod of the application
+// according to the sed script.
+func ModifySourceCode(oc *exutil.CLI, selector labels.Selector, sedScript, file string) error {
```
this feels oddly scoped to me.  why not have this method take a pod or a container, instead of a label, given that it only works if the thing it finds is a single pod and it only actualy operates on the first container in that pod?

mnagy replied:
good point

mnagy replied:
I've looked at it again and it uses the WaitForPods() function which takes a label, the point being that the function will also make sure the pod is running. I could move the waiting-until-pod-is-running logic outside of that function, but I think it will only make the tests look less comprehensible.

mnagy commented:
FYI, Travis failing for unrelated reasons. I checked Jenkins logs and hot deploy tests are succeeding.

bparees commented on test/extended/util/framework.go:
```diff
@@ -452,6 +452,9 @@ func FetchURL(url string, retryTimeout time.Duration) (response string, err erro
 			return false, nil
 		}
+		if r.StatusCode != 200 {
+			return false, fmt.Errorf("Got HTTP response %d, expected 200", r.StatusCode)
```
I think you need to return nil here.  if you return an error, the wait.Poller is going to stop polling.

in the case that an app server returns a 404 until the app is up, you actually do want to keep polling.

mnagy commented:
Addressed @bparees's comment. Additionally, I've added a wait for build, because I discovered that WaitForAnEndpoint() will hang if the build fails. And one minor nit for @jhadvig.

bparees commented:
lgtm.  what is the nit for @jhadvig ?

mfojtik commented:
LGTM as well

mnagy commented:
@bparees adding the postfix parameter here: https://github.com/openshift/origin/blob/7be12f6985226449a7d09f3733542fc5eba9bcac/test/extended/images/helper.go#L46

bparees commented on test/extended/images/helper.go:
```diff
@@ -16,3 +22,36 @@ func GetImageLabels(c client.ImageStreamImageInterface, imageRepoName, imageRef
+
+// CheckPageContains makes a http request for an example application and checks
+// that the result contains given string
+func CheckPageContains(oc *exutil.CLI, endpoint, postfix, contents string) (bool, error) {
```
"postfix"?  path?

bparees commented:
@mnagy needs a squash

jhadvig commented:
LGTM ... lets merge it !

bparees commented:
did the extended tests pass jenkins?

jhadvig commented:
not sure, will re-launch the image ext-tests
[test][extended:core(images)]

mnagy commented:
@bparees you can see in the jenkins console log that Hot Deploy tests passed (search for "hot deploy"). @jhadvig's tag didn't work, maybe he was too late. I've edited my previous comment that didn't include focus.

bparees commented:
@mnagy i'm asked you because i don't want to dig through it since you know the TCs you expect to see run/succeed better than i do.

[merge]

mnagy commented:
@bparees sure, no problem. I just wanted to also give you an easy way to verify I'm not lying, if you wanted to check :)

bparees commented:
@mnagy i trust you..........for now :)

jhadvig commented:
:+1:
